### DISCUSSION TITLE
Example Rhombus/Shrubbery Programs

### DISCUSSION THREAD

**Discussion by samdphillips** [racket/rhombus#186] [Show and tell]
Hi all,

One activity that almost anyone can work on is writing programs in the current Rhombus/Shubbery prototype and sharing their experience is using it.  This feedback can be useful in finding deficiencies in the design, help direct where standard library should go, and show others what a Rhombus future might look like. 

If you are not sure what you could write here's a short list of suggestions:

- Porting the server from ["More: Systems Programming with Racket"](https://docs.racket-lang.org/more/index.html)
- If you have a collection of personal scripts try porting some of them.
- Puzzles and Code Katas

As an example here is an incredibly basic script that I wrote yesterday to install a bunch of packages in a new Racket install I had with some notes.

```
#lang rhombus

/*
1. It's been a few months since I've written and Rhombus.  Looks like the keyword syntax changed since last time.
2. Keyword arguments are ~keyword:, but the `~to` in rename doesn't need a `:`?
3. Don't forget to use the prefix.
4. What is the lexical syntax for symbols?
   - Just use the escape hatch it seems.  Ok that doesn't work.
   - No lexical syntax for `quote`, so use the `symbol(...)` special form.
     - Oh and if it has hyphen use #{}
*/

import:
  pkg:
    rename:
      #{pkg-install-command} ~to pkg_install_command

pkg.pkg_install_command("disposable",
                        "drcomplete",
                        "files-viewer",
                        "http-easy",
                        "memoize",
                        "threading",
                        ~deps: symbol(#{search-auto}))


```

**Comment by jackfirth** [comment]
Here's a list of code puzzle sites to try out:

- [Advent of Code](https://adventofcode.com/)
- [Project Euler](https://projecteuler.net/)
- [Exercism](https://exercism.org/tracks/racket/exercises) (Racket track)

**Comment by samdphillips** [comment]
I have created a [repo](https://github.com/samdphillips/rhombus-examples) and I've updated the [web client](https://github.com/samdphillips/rhombus-examples/tree/main/rhombus-examples-web) example that reuses a few different Racket libraries.

**Comment by samdphillips** [comment]
@soegaard wrote a version of the Norvig Spelling corrector. 
https://github.com/soegaard/rhombus-experiments/blob/main/spelling-corrector/spelling-corrector.rkt

**Comment by samth** [comment]
I've started porting the code in the "More" tutorial, here: https://github.com/samth/rhombus-more/

**Reply by samdphillips** [reply]
The issue you were having with `parameterize` can be fixed by changing the macro to:
```
expr.rule '(parameterize ($a ..., $b ...): $body ... ; ...):
  '(param.parameterize_rhom($a ...,$b ..., begin: $body ...; ...))
```

This is something that has tripped me up with Shrubbery macros a few times.  Things that a Lisper may think of as a single sexp are a short sequence of shrubberies.

**Reply by samth** [reply]
I finished this port, which you can now see here: https://github.com/samth/rhombus-more/blob/main/serve3.rhm

**Comment by samth** [comment]
Another example: a [perceptron](https://github.com/samth/rhombus-more/blob/main/perceptron.rhm) originally by Ken Shan in ISL for the course we're teaching, including a `big-bang` visualization.

**Comment by samdphillips** [comment]
@plane wrote a program using SDL and FFI.

https://github.com/plane/rhombus-test/blob/main/main.rhm
